corgidrp/detector.py

```diff
@@ -520,7 +528,149 @@ def calc_sat_fwc(emgain_arr,fwcpp_arr,fwcem_arr,sat_thresh):
     Returns:
         np.array: lowest full well capacity saturation threshold for frames
     """
-    possible_sat_fwcs_arr = np.append((emgain_arr * fwcpp_arr)[:,np.newaxis], fwcem_arr[:,np.newaxis],axis=1)
-    sat_fwcs = sat_thresh * np.min(possible_sat_fwcs_arr,axis=1)
-
-    return sat_fwcs
+	possible_sat_fwcs_arr = np.append((emgain_arr * fwcpp_arr)[:,np.newaxis], fwcem_arr[:,np.newaxis],axis=1)
+	sat_fwcs = sat_thresh * np.min(possible_sat_fwcs_arr,axis=1)
+
+	return sat_fwcs
+
+def flatfield_residuals(images,planims=None):
+    """Turn this dataset of image frames of neptune or uranus and create matched filters and estimate residuals after 
+     dividing from matched filters
+
+     Args:
+    	images (np.array): 2D array of cropped neptune or uranus image frames
+    	planims (str): name of the planet uranus or neptune
+
+     Returns:
+    	matched_residuals (np.array): residual image frames of neptune or uranus divided by matched filter
+	"""
+    if planims=='neptune' or planims=='uranus' :
+        raster_images = np.array([image[2] for image in images])
+        masks = np.array([image[3] for image in images]) 
+        images_split = np.array(np.split(np.array(raster_images),3))
+        matched_filters = np.array([np.nanmedian(np.stack(images_split[i],2),2) for i in np.arange(0,len(images_split))])
+        matched_filters_smooth = [gauss(matched_filters[i],3) for i in range(len(matched_filters))] 
+        matched_residuals=[];
+        for j in range(len(raster_images)):
+            matched_residuals.append(raster_images[j]/matched_filters_smooth[int(np.floor(j//(len(raster_images)//len(matched_filters_smooth))))])
+    return(matched_residuals)
+	    
+def combine_rasters(residual_images,cent=None,planet=None,band=None):
+    """combine the dataset of residual image frames of neptune or uranus and create flat field 
+    	and associated error
+
+    	Args:
+        	residual_images (np.array): residual image frames of neptune and uranus 
+        	cent (np.array): centroid of the image frames
+        	planet (str):   name of the planet neptune or uranus
+        	band (str):  band of the observation band1 or band4
+        	
+    	Returns:
+        	full_qe (np.array): ideal flat field
+        	full_residuals (np.array): flat field from uranus or neptune images
+        	percens_resel (np.array): percentage error of the flat field
+    """
+    
+    n = 420
+    full_qe=np.random.normal(1,.03,(n,n))
```

Comment:
This is used to estimate the error in the flatfield `std_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel))`
`full_residuals_resel` is the flat field created by combining Neptune or Uranus images
`full_qe_resel` is what the ideal flatfield would be

Reply:
I'd be a little careful about assuming that the ideal flat field is a Gaussian distribution around 1 - there *might* be structure in the flats (e.g., due to deposition on a mask or the EMCCD) - but that's fine, as long as we know the structure of the flat field to high enough degree of accuracy, then that's all that matters

so I'm not sure you can use `full_qe` as you have it currently to measure the success of flight data - you might for simulated data (but Erin did include some structure in their code- they included Hubble-like "measles" for example)

Reply:
so beware that your using `full_qe` as you have it defined might actually *over* estimate the error in the flatfield, since the flat might have structure in it that is not encapsulated by your Gaussian definition of `full_qe`

lastly...I'm not sure we need to actually calculate the error in the flat - we've done that via Erin's simulations - so we know that if we achieve a particular SNR on the images of Uranus/Neptune and apply a similar matched filter method, we should achieve our requirement (with margin)

so I wonder if we should just take this bit out - what do you think, @semaphoreP ?

Reply:
hi @ramya-anche and @semaphoreP - @maxwellmb and I quickly chatted on Slack and we agreed that we won't know the "true" flat field, as flats typically have structure in them, so we cannot use a Gaussian like this to estimate the error in our measured flats

rather, the best we can do is to estimate the measurement uncertainty from the measurements themselves, like the standard deviation of the images per pixel - so we'd effectively measure our measurement uncertainty from the data/measurements themselves

so we recommend removing this Gaussian

instead, we will have the error components already for each frame - so we will have a measurement uncertainty at each dither and then we will have a measurement uncertainty for the matched filter (which we can estimate from the matched filter by taking the stddev across each pixel in the dither), and when you divide one by the other, your error should also propagate - and this error is the one we need to track at the end of the day

Reply:
Hi @rzellem I have modified the `flatfield_residuals` to estimate the errors from individual dithers and matched filters. Can you review and see if this is all right?

Reply:
@semaphoreP - I think we need to compute an "error" to properly track the additional uncertainty that comes in when dividing by our matched filter - because the matched filter itself will have an uncertainty (that we measure by taking the standard deviation along each pixel of all of the "stacked" images of Uranus or Neptune when making the matched filter)

Reply:
Ah I misread your comment, you are also referring to the error in the filter itself, not just the data. I think that's a fine way to do it.

Reply:
not sure I agree with that error propagation, but happy to punt on it for now and revisit it later :)

@ramya-anche - can you please add a "TODO" in your code to note that we need to update the error propagation so we don't forget later?

Reply:
@semaphoreP - in the interest of time, should we punt on this or do we have the time to incorporate it correctly now?

Reply:
Errors are not a requirement. I think we either put one in that's somewhat accurate, or we don't include one at all (as we aren't anticipating flat field to be the dominant noise source). I think probably the easiest path is to not include it for now, because we don't want one that's significantly incorrect. 

The error I quoted is the statistical error of doing a division by a template matched filter (assuming perfect filter). But I would need to look at the details on how you are doing a matched filter to see if it's appropriate.

Reply:
Not including it is basically setting the FlatField err field to all zeros. @ramya-anche, hopefully that should be easy to do.